acquia_lift.admin.inc

```diff
@@ -858,3 +858,866 @@ function acquia_lift_report_conversion() {
+    '#prefix' => '<span class="personalize-add-link-prefix"></span>',
+    '#type' => 'submit',
+    '#value' => t('Add context'),
+    '#submit' => array('acquia_lift_targeting_add_context_submit'),
+    '#ajax' => array(
+      'callback' => 'acquia_lift_targeting_context_ajax_callback',
+      'wrapper' => $main_wrapper_id,
+      'effect' => 'fade',
+    ),
+  );
+  $form['strategies'] = array(
+    '#type' => 'container',
+    '#attributes' => $attributes,
+  );
+
+  // Add radio buttons so the user can select how multiple features for an option
+  // should be treated.
+  $default_strategy = 'OR';
+  $form['strategy'] = array(
+    '#type' => 'select',
+    '#multiple' => FALSE,
+    '#field_prefix' => t('Visitor must have '),
+    '#field_suffix' => t(' of the specified contexts'),
+    '#description' => t('Choose how multiple contexts should be applied to options. Choose "any" if the rule should apply if the user has any of the contexts. Choose "all" if the rule should apply only if the user has all of the contexts.'),
+    '#options' => array(
+      'OR' => 'any',
+      'AND' => 'all'
+    ),
+    '#default_value' => $default_strategy,
+  );
+  $form['submit'] = array(
+    '#type' => 'submit',
+    '#value' => t('Create new audience')
+  );
+  return $form;
+}
+
+/**
+ * Validation callback for the new target audience form.
+ */
+function acquia_lift_new_target_audience_form_validate($form, &$form_state) {
+  $machine_name = personalize_generate_machine_name($form_state['values']['name'], NULL, '-');
+  $agent_name = $form_state['values']['agent'];
+  $option_set = acquia_lift_get_option_set_for_targeting($agent_name);
+  if (isset($option_set->targeting[$machine_name])) {
+    form_set_error('name', t('Please choose a different name for your audience as this one is already taken'));
+  }
+}
+
+/**
+ * Submit callback for the new target audience form.
+ */
+function acquia_lift_new_target_audience_form_submit($form, &$form_state) {
+  $values = $form_state['values'];
+  $contexts = array();
+  // We need to massage the context information as submitted in the form into
+  // an array of contexts that can be consumed by the
+  // acquia_lift_target_audience_save() function.
+  foreach ($values['mapping']['contexts'] as $context_values) {
+    if ($context_values['context'] == '') {
+      continue;
+    }
+    $context_values['match'] = $context_values['value']['match'];
+    $context_values['operator'] = $context_values['value']['operator'];
+    unset($context_values['value'], $context_values['remove']);
+    $contexts[] = $context_values;
+  }
+  if (acquia_lift_target_audience_save($values['name'], $values['agent'], $contexts, $values['strategy'])) {
+    drupal_set_message(t('The new target audience was saved successfully'));
+  }
+  else {
+    drupal_set_message(t('There was a problem saving the new target audience.'), 'error');
+  }
+
+}
+
+/**
+ * Saves a new target audience for an agent.
+ *
+ * @param $label
```

Comment:
OK - so this may be my only substantive comment, but this docblock needs to be filled out ;-)